components/patient-data-extra/medication/api/src/main/java/org/phenotips/data/internal/controller/MedicationController.java

```diff
@@ -111,36 +121,139 @@
-                o.setStringValue(Medication.NAME, m.getName());
-                o.setStringValue(Medication.GENERIC_NAME, m.getGenericName());
-                o.setStringValue(Medication.DOSE, m.getDose());
-                o.setStringValue(Medication.FREQUENCY, m.getFrequency());
-                if (m.getDuration() != null) {
-                    o.setIntValue(DURATION_YEARS, m.getDuration().getYears());
-                    o.setIntValue(DURATION_MONTHS, m.getDuration().getMonths());
+            final XWikiDocument docX = patient.getXDocument();
+            final PatientData<Medication> medications = patient.getData(getName());
+
+            if (medications == null) {
+                if (PatientWritePolicy.REPLACE.equals(policy)) {
+                    docX.removeXObjects(Medication.CLASS_REFERENCE);
                 }
-                if (m.getEffect() != null) {
-                    o.setStringValue(Medication.EFFECT, m.getEffect().toString());
+            } else {
+                if (!medications.isIndexed()) {
+                    this.logger.error(ERROR_MESSAGE_DATA_IN_MEMORY_IN_WRONG_FORMAT);
+                    return;
                 }
-                o.setLargeStringValue(Medication.NOTES, m.getNotes());
+                saveMedications(docX, patient, medications, policy, this.xcontext.get());
             }
-        } catch (Exception ex) {
+        } catch (final Exception ex) {
+            this.logger.error("Failed to save medication data: {}", ex.getMessage(), ex);
+        }
+    }
+
+    /**
+     * Saves {@code medications} data according to the provided {@code policy}.
+     *
+     * @param docX the {@link XWikiDocument}
+     * @param patient the {@link Patient} whose data is being updated
+     * @param medications the medication {@link PatientData data} to update {@code patient} with
+     * @param policy the {@link PatientWritePolicy} according to which data should be saved
+     * @param context the {@link XWikiContext}
+     */
+    private void saveMedications(
+        @Nonnull final XWikiDocument docX,
+        @Nonnull final Patient patient,
+        @Nonnull final PatientData<Medication> medications,
+        @Nonnull final PatientWritePolicy policy,
+        @Nonnull final XWikiContext context)
+    {
+        // A stream of medication objects to save.
+        final Stream<Medication> medsStream;
+        if (PatientWritePolicy.MERGE.equals(policy)) {
+            final PatientData<Medication> storedMedications = load(patient);
+            medsStream = storedMedications != null
+                ? buildMergedMedicationsMap(medications, storedMedications).entrySet().stream()
+                    .flatMap(mapEntry -> mapEntry.getValue().stream())
+                : StreamSupport.stream(medications.spliterator(), false)
+                    .filter(Objects::nonNull);
+            docX.removeXObjects(Medication.CLASS_REFERENCE);
```

Comment:
if we anyways `docX.removeXObjects(Medication.CLASS_REFERENCE);` in both `if` and `else`, may be we can pool it out and place before `if`?